Some improvements to loading the session with --prompt-cache

KerfuffleV2 opened the pull request:
### `--seed` is ignored when loading session

Currently the `--seed` parameter is ignored when loading the prompt. However, a very common use case would be to save a prompt and then try several attempts at generation with different seeds.

The pull includes a simple change that just sets the RNG if specified. Two small notes:

1. There isn't a way to tell if `-seed` was actually actually specified as far as I know, only that it's not the default `-1` value. So `--seed -1` is the same as not including it: it won't override the cached seed.
2. The RNG won't be in the same state as if the seed had been specified originally. I.E. If you generate the cached prompt using `--seed 123` and then load it with `-seed 123` the subsequent tokens will not match. I don't think there's an easy way around this. It's not 100% ideal but still a lot better than just completely ignoring the parameter with no warning.

### Blank prompt overwrites cached one

When loading a cached prompt from a session, you have to specify the prompt again. Even worse, if you forget to enter a prompt you'll get your cached prompt overwritten by the blank one because of course it has low similarity.

This pull changes that behavior to simply use the tokens from the saved session if `params.prompt` is empty (in other words not set via `--prompt` or `--file`). 


Closes #1439

KerfuffleV2 committed:
Some improvements to loading the session with --prompt-cache

1. Currently the --seed parameter is ignored when loading the prompt. However, a very common use case would be to save a prompt and then try several attempts at generation with different seeds.
2. When loading a cached prompt from a session, you have to specify the prompt again. Even worse, if you forget to enter a prompt you'll get your cached prompt overwritten by the blank one.

KerfuffleV2 committed:
Apply clang suggestions.

ejones commented:
LGTM! Tested both features, works great!

DannyDaemonic commented:
I like the seed fix. Personally, I would prefer it always overrides the seed with a random one if it's set to -1. If we want to keep the seed the same, we could pass it in each time. I don't see a need to keep the seed, but if you can think of a use case, perhaps we could check if seed was set to something like -2 and only then keep the original seed?

Also, currently, prompt cache treats text given during `--interactive-first` as the prompt when one isn't specified. I've found this to be very useful because that's how I usually feed my prompt in. This PR seems to break that use case.

KerfuffleV2 commented:
@DannyDaemonic 

> Personally, I would prefer it always overrides the seed with a random one if it's set to -1.

I agree. The issue is `-1` is the default value for `seed` in the `gpt_params` struct. So there isn't currently a way to differentiate between `--seed` being unspecified and `--seed -1` being supplied.

The only way to fix that would be store something like a `seed_was_set` bool in the params also (or use some kind of container for values that could indicate if it was default or user supplied). All of that sounds too complicated for the benefit, when you can just use `--seed -2` or any other negative value if you want to make sure the seed is random.

> Also, currently, prompt cache treats text given during `--interactive-first` as the prompt when one isn't specified. [...] This PR seems to break that use case.

I will check that and get back to you. Breaking any current behavior definitely is not intended.

KerfuffleV2 committed:
Use existing session behavior when in instruct or interact first mode

KerfuffleV2 commented:
@DannyDaemonic 

> Still, if you think the seed should only be overridden when explicitly set

I don't want to change the existing behavior too much. I'm a little confused though, I think it already works the way you want without any further changes.

With this pull, if you specify _any_ negative seed _except_ for `-1`, that will make it use a random seed even when loading a session. So you can just use `--seed -2` when loading a session and get new results each time.

***

Also, can you please check and see if the change I just pushed fixes your interactive mode issue? (The behavior should now be identical to the master branch behavior.)

DannyDaemonic reviewed (commented):
I just tested the PR and it does preserve the use with `--interactive-first`. However, I found a bug related to your random seed code.

---

> I don't want to change the existing behavior too much. I'm a little confused though, I think it already works the way you want without any further changes.

> With this pull, if you specify any negative seed except for -1, that will make it use a random seed even when loading a session.

Sorry, I'm talking about two separate solutions; maybe that's where the confusion is coming from.

The first possible solution is to just *always* override the seed. Although your PR says otherwise, this is actually what your code currently does. Since lines [156 to 158](https://github.com/ggerganov/llama.cpp/blob/41530b67c14758e3ad43d6be4231e865ff72122b/examples/main/main.cpp#L156-L158) never do anything, they should just be removed. Simple solution.

This will mean even without `--seed` specified, we update the seed. I think the seed not being updated was simply an oversight, not a design decision. To me, this is the expected behavior. Again, the argument being, it's only a prompt cache, not a full session cache.

The other suggestion is that `-2` is the seed you ignore instead of `-1`. This would maintain the existing behavior of `--seed -1`. To accomplish this, you could set `params.seed = -2;` before `gpt_params_parse` is called on [line 54](https://github.com/ggerganov/llama.cpp/blob/41530b67c14758e3ad43d6be4231e865ff72122b/examples/main/main.cpp#LL54C9-L54C25). This is a bit messier to me, and I've seen people ask why the seed isn't being randomized anymore (which is probably not being done with `--seed -1`). I also think it's a bit early to worry that people are depending on the seed not changing.

But either way, both of these solutions work as expected when the user specifies `--seed -1`.

aleksusklim commented:
What happens when a prompt stored in `--prompt-cache` is "longer" than the current one?

I want to make a wrapper around `main.exe`, but this behavior looks strange and buggy. I don't know whether it is a bug or not, but it looks very-very confusing. I may post it as a separate issue if that's really a bug that you cannot solve in this PR.

So, here are my steps: (on version ee9654138ab0ae5f138f4abddf56ca234ea3c352 and [this model](https://huggingface.co/TheBloke/WizardLM-7B-uncensored-GGML))

`main.exe -t 6 -m WizardLM-7B-uncensored.ggml.q5_1.bin -c 512 --temp 0 --repeat_penalty 1.2 --prompt-cache cache1 -f in.txt >out.txt`
(note the cache file and zero temperature)

My prompt inside `in.txt` is this:

```
### Instruction:

Describe your last dream in a few words.

### Response:

My
```

The model outputs this text inside `out.txt` (with an extra space before the first line, but let's assume I stripped it manually):

```
### Instruction:

Describe your last dream in a few words.

### Response:

My last dream was about being chased by a group of people through a forest, but I eventually found safety inside a church.
```

It also creates prompt-cache file `cache1` with size around 13 Mb, and writes to stderr:
```
main: attempting to load saved session from 'cache1'
main: session file does not exist, will create
```

If I repeat the same command as-is – it recreates the same text and does not update the cache file, with stderr being:
```
main: attempting to load saved session from 'cache1'
main: loaded a session with prompt size of 26 tokens
main: session file has exact match for prompt!
```

Then I copy "cache1" file to `cache2` file. I also put the resulting text back into `in.txt`, this time cutting it after the last comma, so it becomes:

```
### Instruction:

Describe your last dream in a few words.

### Response:

My last dream was about being chased by a group of people through a forest,
```

Then I run, pointing to `cache2`: (which is the same as cache1 for now)
`main.exe -t 6 -m WizardLM-7B-uncensored.ggml.q5_1.bin -c 512 --temp 0 --repeat_penalty 1.2 --prompt-cache cache2 -f in.txt >out.txt`

It gives the exact same line, continuing with `but I eventually found safety inside a church.` and stopping as before.
But this time, `cache2` is updated to 21 Mb. So far so good!
Its stderr said:
```
main: attempting to load saved session from 'cache2'
main: loaded a session with prompt size of 26 tokens
main: session file matches 26 / 42 tokens of prompt
```

Finally, I copy cache2 to `cache3` and cut the prompt back to `My` just as my very first in.txt contents.
I run, pointing to cache3, and get the following:

```
### Instruction:

Describe your last dream in a few words.

### Response:

My butterfly had wings of fire and flew through the night sky while I chased after it, trying to catch it so that we could be together forever.
```

File `cache3` stays binary equal to cache2, and the program outputs to stderr this lines among the others:
```
main: attempting to load saved session from 'cache3'
main: loaded a session with prompt size of 42 tokens
main: session file has exact match for prompt!
```

What just happened? For me, it looks like the program compared for the head of the prompt discarding its cached tail; then it assumed that the cache is valid (despite it is not) and continued the generation from the cache, ending up in a broken state.

My questions:
1) Is it possible to "go back" for any amounts of tokens in the cache? Can we cut the prompt in arbitrary place and keep the cache valid? Or is it only valid "after the final token", so we can go forth but cannot start earlier?
2) If we cannot go back, is it possible to "restore text" from the cache and output it explicitly? So the program would show, what we have in the cache, and what is causing its weird answers.
3) If yes, can we store "checkpoints" in the cache with full intermediate state, so that in my example after seeing "My" at the end – it would use the older state that it already had; or if I'll stop at "My last dream" then it continues up to "was about being chased by a group of people through a forest," (since that state was also recorded) and not break there?
4) If yes, why can't we "discard all of extra states" upon seeking, for example, "My last dream was dark" – backtrack to "My" (which is stored), then evaluate just "last dream was dark" and generate from here? Just as If I put a copy of an older cache myself.
5) Can't it at least throw an error with mismatching cache (or discard it and regenerate from scratch) when it detects that the prompt is different?
6) Can we specify "do not update cache" command-line option? Or "write new cache to the specified file" (which will also accept NUL to skip it) so it will not destroy the previous state? Currently my only option is to copy the cache file each time before invoking the program, which delays startup because of OS filesystem cache eviction when the model barely fits in RAM already.

And a few minor things:
– Why not to run the evaluation with `--n-predict 0` if I only want to cache the prompt? I have to specify `--n-predict 1` to just cache the prompt, but that prevents me to use (probably reparsed?) output file since it will contain an extra word at the end. I have to use the initial file.
– Why to print an extra space at the beginning of stdout? If I swap files (giving out.txt to in.txt) later, it will grow each time (two spaces, three spaces…) and most likely destroys the cache. I had to strip that manually, but this feels odd.

KerfuffleV2 commented:
@aleksusklim

If I'm reading your post correctly, it sounds like you haven't even tried this pull request? I'd like to stay on topic here since it's confusing to talk about other stuff unrelated to my changes. I initially thought you were saying my pull introduced new issues.

If you're just asking general questions about prompt caching behavior then creating a discussion in the Q&A section is probably most appropriate: https://github.com/ggerganov/llama.cpp/discussions/categories/q-a

You could possibly create an issue instead but Q&A discussion makes the most sense to me. 

If you want to @ me when you do that, I can probably answer some of your questions.

KerfuffleV2 committed:
Use the initial value of params.seed to determine if user supplied seed

Display some helpful information to the user when loading a session to make it clear when the seed applies or not.

KerfuffleV2 commented:
@DannyDaemonic Thanks for taking the time to test and catching that problem! Since `llama_load_from_file` already has that logic I wasn't looking for it in other places as well.

> To me, this is the expected behavior. Again, the argument being, it's only a prompt cache, not a full session cache.

The thing is, it actually saves and loads the RNG state there. So apparently someone though it was important to preserve that state. I'm reluctant to change that behavior in something that's mainly supposed to be a bug fix.

***

I've made some further changes. Can you please test again? Hopefully I actually got it right this time - so much for a simple fix!

I don't love the approach of having to save the seed parameter before it gets overwritten but I'm not sure there's really a better way without making significant changes.

This is how the RNG/session loading interaction is supposed to work with the latest changes:

* If `--seed` isn't specified (or `--seed -1` is used) then the saved RNG state in the session file will be used. I also added a message to indicate this is happening (because application prints out the seed no matter what, even if it's not actually used)
* If `--seed` is specified with any positive value then the session RNG state is ignored. Generation starts with a RNG initialized to the specified seed. It will also print out a message informing the user this occurred.
* If `--seed` is specified with any negative value other than `-1` then same as the above aside from generating a new seed.

aleksusklim commented:
> it sounds like you haven't even tried this pull request

No, I did not. I don't see any point to "fix things" in seed with `--prompt-cache` if this prompt cache is broken in other ways that prevent to use it altogether.

KerfuffleV2 commented:
@aleksusklim 

> No, I did not. I don't see any point to "fix things" in seed

Okay, well, hopefully you can appreciate how it would be confusing to start talking about different issues in a pull request aiming to fix some other problem.

Also, there's obviously a benefit to fixing existing problems even if it result in something that works perfectly for every use case. At the least, you're closer to a state where it will do what you want. Fixes and improvements are an incremental process, it's not all or nothing.

> if this prompt cache is broken in other ways that prevent to use it altogether.

Maybe it's broken for your specific use case, but that doesn't mean it's broken/useless altogether. From your description, it seems like you're doing something unusual and pretty specific.

Anyway, my advise is to create a discussion. Like I said, if you @ me I will try to help you out by answering some of your questions there.

DannyDaemonic commented:
> The thing is, it actually saves and loads the RNG state there. So apparently someone though it was important to preserve that state. I'm reluctant to change that behavior in something that's mainly supposed to be a bug fix.

This is using a feature in the llama.cpp API that stores and loads the entire state. This is so you can do things such as save states for a web server where you're having conversations with many people at once. Everything is ready to go from the stored state so you can just swap in whatever state you need and evaluate the next response.

In this case however, the state is only being used as a --prompt-cache. Restoring things blindly is also causing issue #1585. It's probably also causing other settings besides `--seed` to be ignored as well. Seed is just the most obvious because everytime you load it, you get the same response.

However, now that I'm looking at it. I think the rng being ignored is a bug in itself:
```C++
    ctx->rng = std::mt19937(params.seed);
    ctx->logits_all = params.logits_all;

    ggml_type memory_type = params.f16_kv ? GGML_TYPE_F16 : GGML_TYPE_F32;

    if (!llama_model_load(path_model, *ctx, params.n_ctx, params.n_gpu_layers, memory_type,
                          params.use_mmap, params.use_mlock, params.vocab_only,
                          params.progress_callback, params.progress_callback_user_data)) {
        fprintf(stderr, "%s: failed to load model\n", __func__);
        llama_free(ctx);
        return nullptr;
    }
```

The correct place to fix this is probably in llama.cpp itself.

aleksusklim commented:
> At least, that person hasn't show an actual problem there yet.

CharacterAI chat: https://github.com/Cohee1207/SillyTavern/blob/999a94718df39227fc5ea2bc80006351f43c5a88/public/instruct/WizardLM.json#L3

```
Write AI's next reply in a fictional roleplay chat between User and AI.

### Instruction:

User: Hello, I …
AI: Great! I think…
User: What if I just…

### Response:

AI: 
```

Then it will print something like:
`In that case, we will…`

And I need to craft this one:

```
Write AI's next reply in a fictional roleplay chat between User and AI.

### Instruction:

User: Hello, I …
AI: Great! I think…
User: What if I just…
AI: In that case, we will…
User: <NEXT USER INPUT GOES HERE>

### Response:

AI: 
```

This is not "adding to the end", it should be pasted after "User:" but before "### Response:", which means that I need to restore session from there somehow. For example, but cutting after last user input and fake-run just to save session to a separate file; then put Response-tail template and save another session derived from it; then regenerate at different seeds (if that would be possible of course!) or settings until I satisfied with the answer; then add it to the prompt and change the initial session the same way.
(Otherwise I will have to feed it with the same chat history over and over again, which quickly becomes the bottleneck).

That requires my wrapper to have FULL control over user prompts, and tracking which one related to each session copy.
But I see that main.exe tries to recover from incomplete sessions on its own. So I'm not sure, on which behavior I should rely?

The best behavior would be if I won't even care about sessions, just give it a file and then work with the prompt ONLY, so the main.exe will decide: will it use the cache, update it, discard it – without any effects on the final result, just as it was generated from scratch.

Currently, when given shorter prompt than in session – I get _irrelevant_ replies, not just randomized! Roleplay just breaks if I change prompt arbitrary when using the same session file without copy-swapping it manually.

> I think the only way around that would be to store the RNG state after every token had been generated.

Aren't the "state" (context) is stored? Cache files are large, I presume they have all internal variable memory of the model.

DannyDaemonic commented:
> > Restoring things blindly is also causing issue https://github.com/ggerganov/llama.cpp/issues/1585.

> I'm not fully convinced that's in issue. At least, that person hasn't show an actual problem there yet.

I think I know what he's talking about. I've seen similar things myself. Try this:
```
./main -m /path/to/llama/bin --prompt-cache Z.cache -p "Here's a funny joke. No, wait, here's a bunch of Zs: Z Z Z Z Z Z Z Z Z Z"
```
Then try this:
```
./main -m /path/to/llama/bin --prompt-cache Z.cache -p "Here's a funny joke."
```

The joke will start with a `Z` every time. Something is wrong but I don't know what. I think something is being restored that shouldn't be. Again, because we're restoring the entire state, but we're just trying to use the prompt cache. Perhaps the logits are not being reevaluated for some reason.

> That couldn't fix the problem by itself anyway, since the main example is overwriting the seed param on its own.

The main example doesn't call `llama_set_rng_seed` until your patch. It's set by the code in the function I posted earlier. Upon looking more closely, it is probably restored intentionally. I just thought it might be a bug because it's initialized and then overwritten when you'd typically have a different path.

I do think setting it here in main is a good idea.

DannyDaemonic commented:
> I don't really think it's related to saving the RNG state though. With my patch and overriding the RNG and reseeding it, I still see that behavior.

Absolutely not. My point was just because something is restored by the state doesn't mean we need to use it for our prompt cache.

> Are you saying you think the current approach I'm using is okay after all?
>
> What's your opinion of the current state of this pull request. Do you think anything needs to change before merging?

I'm still leaning towards always replacing the seed, no matter what.  (Again, it's only a cache.)

At the very least, we should honor `--seed -1` since that's what's used in the --help, in the examples, and in the docs. To always honor `-1` you can just set `params.seed` it to `-2` before `gpt_params_parse` on line 54 and test for `-2` instead of `-1` as the exception.

DannyDaemonic commented:
@KerfuffleV2 
> If anyone with authority (which might be you?) tells me it should be a certain way then I'm happy to accept it and do it that way.

I don't know if there's really any kind of hierarchy around here, other than ggerganov has the ultimate say since he's the mastermind behind it all. His philosophy seems to be: have fun, don't be afraid of breaking changes, and iterate. I think he generally relies on everyone to discuss changes and come up with reasonable solutions (while avoiding drama).

I wouldn't be too worried about changing the behavior yet. All of this is very new, and as ejones said, "nothing about prompt (or event state restoration) is particularly sacred at this point."

I think it's also the simpler solution; just overwrite it every time and to treat the `--prompt-cache` as strictly prompt cache. This means if ejones does come up with a more efficient ("surgical") state save format then he doesn't have to hold onto the `mersenne_twister_engine` data from the RNG.

That said, if *you* are depending on this *feature* as is, then let's go with option B and leave the seed as unchanged by default.

> The only thing I'd change is using something a user would be less likely to select than `-2`.

At the start of this I almost suggested using `MIN_INT` myself, but then I was thinking `-2` would be an easy number to remember. Either is fine really, but if we go this route I'll also probably have to update the documents under the `--prompt-cache` write-up in the [README](https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md) with the number as a "Here be dragons" type warning. (My thinking was, if it's a hard number to remember then the warning doesn't do as much good.)

ggerganov commented:
> His philosophy seems to be: have fun, don't be afraid of breaking changes, and iterate. I think he generally relies on everyone to discuss changes and come up with reasonable solutions

Yes, that's well said

I am paying less attention to the changes in the examples and trust that you will figure out what's right.
As long as the examples demonstrate the basic functionality of the model, everything else is a bonus.
Breaking stuff in the examples is fine, since that part of the code is not directly used into 3rd party projects and so we can afford to experiment

x4080 commented:
waiting for this

KerfuffleV2 committed:
Always set RNG seed when restoring cached prompt in main example.

Add a note in the main example README about how restoring a prompt doesn't imply restoring the exact session state.

KerfuffleV2 commented:
Thanks everyone that replied, appreciate the information!

@DannyDaemonic 

> I think it's also the simpler solution; just overwrite it every time

I'm fully convinced! I made that change and also added a little note to the README mentioning that restoring a cached prompt doesn't mean you get the same sequence of tokens as the original generation, even when specifying the same seed.

> That said, if you are depending on this feature as is

Nope, I only cared about being able to try multiple seeds from a cached prompt without having to repeatedly specify it.

> have fun, don't be afraid of breaking changes

As a pessimistic, risk-averse type of person I fear that would require brain surgery...

KerfuffleV2 commented on examples/main/README.md:
```diff
@@ -272,7 +272,7 @@ These options help improve the performance and memory usage of the LLaMA models.
 ### Prompt Caching
 
--   `--prompt-cache FNAME`: Specify a file to cache the model state after the initial prompt. This can significantly speed up the startup time when you're using longer prompts. The file is created during the first run and is reused and updated in subsequent runs.
+-   `--prompt-cache FNAME`: Specify a file to cache the model state after the initial prompt. This can significantly speed up the startup time when you're using longer prompts. The file is created during the first run and is reused and updated in subsequent runs. **Note**: Restoring a cached prompt does not imply restoring the exact state of the session at the point it was saved. So even when spe ...
```
I thought about that, but we'd also have to add that this _only_ applies with the same generation settings and same exact build. Anything that changes how sampling occurs or even causes one less or one more random number to be generated at some point will throw the whole thing out of whack.

Also I think some BLAS implementations also can cause output to be non-deterministic even when meeting the other requirements (at least whenever I run the main example with cuBLAS I get a warning about when specifying a seed).

So it kind of seemed like it would just confuse the user to explain all that, and the alternative is to not tell them the whole truth. Maybe one way to deal with it would be just phrase it in a vague way like "generally speaking, if you use same seed and prompt you'll get the same result when restoring a session but changing settings or using a different version can affect this". 

What do you think?

DannyDaemonic replied:
Yeah, I see what you're saying; I forgot that some BLAS implementations can cause things to diverge.

The docs don't mention it anywhere at this point, not even under `--seed`. I think perhaps it's a caveat that's just understood by those BLAS users?

It might be worth adding a section on BLAS to the README that explains these caveats. (Not in this patch.)

DannyDaemonic replied:
I do believe I can edit the PR, but I also think you can just click accept and it'll merge the changes. That said, if f you're just trying to keep your hands clean, I'll approve it as is. 😄

KerfuffleV2 replied:
> That said, if f you're just trying to keep your hands clean

I guess you could call it that. :) I wouldn't really add that myself because it's (at least technically) incorrect. Like I said though, I'm okay with _you_ committing it/editing this pull/whatever.

The main thing I care about is the changes to the actual code.
